standalone-metastore/metastore-catalog/src/main/java/org/apache/iceberg/HiveCachingCatalog.java

```diff
@@ -0,0 +1,333 @@
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+
+/**
+ * Class that wraps an Iceberg Catalog to cache tables.
+ * Initial code in:
+ * https://github.com/apache/iceberg/blob/1.3.x/core/src/main/java/org/apache/iceberg/CachingCatalog.java
+ * Main difference is the SupportsNamespace.
+ *
+ * <p>See {@link CatalogProperties#CACHE_EXPIRATION_INTERVAL_MS} for more details regarding special
+ * values for {@code expirationIntervalMillis}.
+ * @param <CATALOG> the catalog class
+ */
+public class HiveCachingCatalog<CATALOG extends Catalog & SupportsNamespaces> implements Catalog, SupportsNamespaces {
+  private static final Logger LOG = LoggerFactory.getLogger(HiveCachingCatalog.class);
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final long expirationMs;
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final Cache<TableIdentifier, Table> tableCache;
+  private final CATALOG catalog;
+
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected HiveCachingCatalog(CATALOG catalog, long expiration) {
+    Preconditions.checkArgument(expiration > 0,
+        "When %s is <= 0, the catalog cache should be disabled.",
+        CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS);
+    this.catalog = catalog;
+    this.expirationMs = expiration;
+    this.tableCache = createTableCache(Ticker.systemTicker());
+  }
+
+  public CATALOG unwrap() {
+    return catalog;
+  }
+
+  public static <C extends Catalog & SupportsNamespaces>
+  HiveCachingCatalog<C> wrap(C catalog, long expirationIntervalMillis) {
+    return new HiveCachingCatalog(catalog, expirationIntervalMillis);
+  }
+
+
+  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
+    Caffeine<Object, Object> cacheBuilder = Caffeine.newBuilder().softValues();
+    if (expirationMs > 0) {
+      return cacheBuilder
+          .removalListener(new MetadataTableInvalidatingRemovalListener())
+          .executor(Runnable::run) // Makes the callbacks to removal listener synchronous
+          .expireAfterAccess(Duration.ofMillis(expirationMs))
+          .ticker(ticker)
+          .build();
+    }
+    return cacheBuilder.build();
+  }
+
+  /**
+   * Eventual canonical transformation.
+   * @param tableIdentifier the identifier
+   * @return the canonical representation
+   */
+  protected TableIdentifier canonicalizeIdentifier(TableIdentifier tableIdentifier) {
```

Comment:
It does nothing just gives back the parameter. What is the purpose of this method?

Reply:
It looks like Iceberg has something similar. But it has functionality: https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/CachingCatalog.java#L121